### PR TITLE
add generic and typed watch to KubernetesObjectApi

### PR DESCRIPTION
**Pull request by schrodit**
Follow up of https://github.com/kubernetes-client/javascript/pull/1705#issuecomment-2148569436

### CODE REVIEW

File: src/object.ts [hunk 1]
```diff
@@ -44,6 +45,60 @@ enum KubernetesPatchStrategies {
...
+    object: T;
+};
+
+export type WatchCallback<T extends KubernetesObject | KubernetesObject> = (
+    phase: KubernetesEventType,
+    apiObj: T,
+    watchObj?: WatchObject<T>,
+) => void;
+
+export type WatchOptions = {
```
Comment (schrodit) [inline src/object.ts:79]: I think those are the only 2 options that make sense for watches

File: src/object.ts [hunk 2]
```diff
@@ -474,6 +532,64 @@ export class KubernetesObjectApi extends ApisApi {
...
+        done,
+    }: {
+        resource: {
+            apiVersion: string;
+            kind: string;
+            namespace?: string;
+        };
+        options?: WatchOptions;
+        callback: WatchCallback<T>;
+        done: (err: unknown) => void;
```
Comment (schrodit) [inline src/object.ts:562]: might make sense to also have done as optional callback

**Comment by brendandburns** [comment]
Looks like CI/CD is failing for some reason.

**Comment by schrodit** [comment]
> This looks good to me. Any chance that you can add unit tests? If not I'll do it as a follow up.

added some tests

File: src/util.ts [hunk 3]
```diff
@@ -162,3 +162,35 @@ export async function generateTmpFileName(): Promise<string> {
...
 }
+
+/**
+ * If you don't want to or can't await a promise right await and need to store
+ * it in a variable to await it later, wrap it with this function.
+ * Otherwise, if the promise rejects before you await it, it will cause an
+ * unhandled promise rejection which can cause process termination (in node).
+ */
+export const awaitLater = <T, P extends Promise<T>>(p: P): P => {
+    p.catch(() => {});
```
Comment (mstruebing) [inline src/util.ts:173]:
```suggestion
    p.catch(() => {}); // tslint:disable-line:no-empty
```

**Comment by schrodit** [comment]
@mstruebing can you take another look?

Do we have to port it to the release-1 branch?

**Comment by brendandburns** [comment]
/lgtm
/approve

Thanks for the PR. It would be appreciated (but not required) to port this to the 1.x branch also.

Soon the 0.x branch will be deprecated and will be replaced by the 1.x branch.

**Pull request merged**